micro-rdk/src/common/conn/viam.rs

```diff
@@ -0,0 +1,1723 @@
+            executor,
+            storage: self.storage,
+            http2_server: self.http2_server,
+            webrtc_configuration: self.webrtc_configuration,
+            http2_connector,
+            mdns: RefCell::new(mdns),
+            component_registry: self.component_registry,
+            provisioning_info: self.provisioning_info,
+            http2_server_insecure: self.http2_server_insecure,
+            http2_server_port: self.http2_server_port,
+            app_uri: self.initial_app_uri,
+            wifi_manager: Rc::new(self.wifi_manager),
+            app_client_tasks: self.app_client_tasks,
+            max_concurrent_connections: self.max_concurrent_connections,
+            network: None,
+        }
+    }
+}
+pub struct ViamServer<Storage, C, M> {
+    executor: Executor,
+    storage: Storage,
+    http2_server: HTTP2Server,
+    webrtc_configuration: WebRtcListener,
+    http2_connector: C,
+    provisioning_info: ProvisioningInfo,
+    mdns: RefCell<M>,
+    component_registry: Box<ComponentRegistry>,
+    http2_server_insecure: bool,
+    http2_server_port: u16,
+    app_uri: Uri,
+    wifi_manager: Rc<Option<Box<dyn WifiManager>>>,
+    app_client_tasks: Vec<Box<dyn PeriodicAppClientTask>>,
+    max_concurrent_connections: u32,
+    network: Option<Box<dyn Network>>,
+}
+impl<Storage, C, M> ViamServer<Storage, C, M>
+where
+    Storage: ViamServerStorage,
+    <Storage as RobotConfigurationStorage>::Error: Debug,
+    ServerError: From<<Storage as RobotConfigurationStorage>::Error>,
+    C: ViamH2Connector + 'static,
+    M: Mdns,
+{
+    pub fn run_forever(&mut self) -> ! {
+        let exec = self.executor.clone();
+        exec.block_on(Box::pin(self.run()));
+    }
+
+    pub(crate) async fn run(&mut self) -> ! {
+        // The first step is to check whether or not credentials are populated in
+        // storage. If not, we should go straight to provisioning.
+        //
+        // truth table for 2nd check
+        // +------------------+------------------+------------------+
+        //| Has WifiManager  | Has Credential   | Should provision |
+        //+------------------+------------------+------------------+
+        //| false            | x                | false            |
+        //| true             | false            | true             |
+        //| true             | true             | false            |
+        //+------------------+------------------+------------------+
+        if !self.storage.has_robot_credentials()
+            || self
+                .wifi_manager
+                .as_ref()
+                .as_ref()
+                .is_some_and(|_| !self.storage.has_wifi_credentials())
+        {
+            self.provision().await;
+        }
+
+        // Since provisioning was run and completed, credentials are properly populated
+        // if wifi manager is configured loop forever until wifi is connected
+        if let Some(wifi) = self.wifi_manager.as_ref().as_ref() {
```

Comment:
Should it be possible to create the robot even if the local wifi is unreachable? You still could capture data. If you find later that the robot is no longer valid once you connect to wifi and then app you would destroy it.